File: 05-dynamic-programming/01_climbing_stairs/climbing_stairs.cpp
/**
 * Problem: Climbing Stairs
 * LeetCode: #70 (Easy)
 * 
 * Time Complexity: O(N)
 * Space Complexity: O(1)
 */

#include <iostream>

int climbStairs(int n) {
    if (n <= 2) return n;
    int prev2 = 1, prev1 = 2;

    for (int i = 3; i <= n; ++i) {
        int curr = prev1 + prev2;
        prev2 = prev1;
        prev1 = curr;
    }
    return prev1;
}

int main() {
    int n = 6;
    std::cout << "[C++] Climbing Stairs Test\n";
    std::cout << "Ways to climb " << n << " stairs: " << climbStairs(n) << " (Expected: 13)\n";
    return 0;
}

File: 05-dynamic-programming/01_climbing_stairs/climbing_stairs.py
"""
Problem: Climbing Stairs
LeetCode: #70 (Easy)

Time Complexity: O(N)
Space Complexity: O(1)
"""

def climb_stairs(n: int) -> int:
    if n <= 2:
        return n
    prev2, prev1 = 1, 2
    for _ in range(3, n + 1):
        prev2, prev1 = prev1, prev2 + prev1
    return prev1

if __name__ == "__main__":
    print("[Python] Climbing Stairs Test")
    tests = [(2, 2), (3, 3), (4, 5), (5, 8)]
    for n, expected in tests:
        res = climb_stairs(n)
        print(f"n: {n} -> Ways: {res} | Expected: {expected}")
        assert res == expected
    print("Climbing stairs tests passed!")
